aptos-move/mvhashmap/src/lib.rs

```diff
@@ -207,30 +240,37 @@ impl<K: Hash + Clone + Eq, V: TransactionWrite> MVHashMap<K, V> {
 
                             if maybe_value.is_none() {
                                 // Resolve to the write if the WriteOp was deletion
-                                // (MoveVM will observe 'deletion').
+                                // (MoveVM will observe 'deletion'). This takes precedence
+                                // over any speculative delta accumulation errors on top.
                                 let write_version = (*idx, *incarnation);
                                 return Ok(Version(write_version, data.clone()));
                             }
-
-                            return accumulator
-                                .apply_to(maybe_value.unwrap().into())
-                                .map_err(|_| DeltaApplicationFailure)
-                                .map(|result| Resolved(result));
+                            return accumulator.map_err(|_| DeltaApplicationFailure).and_then(
+                                |a| {
+                                    // Apply accumulated delta to resolve the aggregator value.
+                                    a.apply_to(maybe_value.unwrap().into())
+                                        .map(|result| Resolved(result))
+                                        .map_err(|_| DeltaApplicationFailure)
+                                },
+                            );
                         },
                         (EntryCell::Delta(delta), Some(accumulator)) => {
-                            // Read hit a delta during traversing the
-                            // block and aggregating other deltas. Merge
-                            // two deltas together. If Delta application
-                            // fails, we return a corresponding error, so that
-                            // the speculative execution can also fail.
-                            accumulator
-                                .merge_onto(*delta)
-                                .map_err(|_| DeltaApplicationFailure)?;
+                            *accumulator = accumulator.and_then(|mut a| {
+                                // Read hit a delta during traversing the block and aggregating
+                                // other deltas. Merge two deltas together. If Delta application
+                                // fails, we record an error, but continue processing (to e.g.
+                                // account for the case when the aggregator was deleted).
```

Comment:
so this makes the deletion observable in case of overflow?

Reply:
yeah, it should keep going as err() instead of short circuiting, and then observing deletion should take precedence.